xc/device.ml

```diff
@@ -1912,6 +1912,29 @@ let stop ~xs ~qemu_domid domid  =
     stop_vgpu ();
     stop_qemu ()
 
+let is_hvm_linux domid =
+	let _XEN_IOPORT_LINUX_PRODNUM = 3 in (* from Linux include/xen/platform_pci.h *)
+	try
+		let open Qmp in
+		match qmp_write_and_read domid (Command (None, Query_xen_platform_pv_driver_info)) with
+		| Some (Success (None, Xen_platform_pv_driver_info { product_num; build_num })) ->
+			((product_num = _XEN_IOPORT_LINUX_PRODNUM) && (build_num <= 0xff))
+		| Some x ->
+			error "Unexpected QMP response: %s" (Qmp.string_of_message x);
+			false
+		| None -> false
+	with e ->
+		error "Exception attempting to obtain Linux product_id and build_number: %s" (Printexc.to_string e);
+		false
+
+let notify_unplug ~xs domid =
+	if (is_upstream_qemu domid) && (is_hvm_linux domid) then
+	begin
+		let write_local_domain prefix x = xs.Xs.write (Printf.sprintf "/local/domain/%d/%s%s" domid prefix x) "1" in
+		List.iter (write_local_domain "control/feature-") ["suspend"; "shutdown"; "vcpu-hotplug"];
```

Comment:
There is also `xs.Xs.writev` for writing a list of key/value pairs to a xenstore tree.